test/publisher_subscriber/PublisherSubscriberInteraction.cpp

```diff
@@ -0,0 +1,144 @@
+{
+public:
+    const uint16_t AGENT_PORT = 2018;
+
+    PublisherSubscriberInteraction()
+    : transport_(std::get<0>(GetParam()))
+    , publisher_(std::get<1>(GetParam()), 8)
+    , subscriber_(std::get<1>(GetParam()), 8)
+    {
+        init_agent(AGENT_PORT);
+    }
+
+    ~PublisherSubscriberInteraction()
+    {
+        agent_->stop();
+    }
+
+    void SetUp() override
+    {
+        ASSERT_NO_FATAL_FAILURE(publisher_.init_transport(transport_, "127.0.0.1", AGENT_PORT));
+        ASSERT_NO_FATAL_FAILURE(subscriber_.init_transport(transport_, "127.0.0.1", AGENT_PORT));
+
+        ASSERT_NO_FATAL_FAILURE(publisher_.create_entities_xml(1, 0x80, UXR_STATUS_OK, 0));
+        ASSERT_NO_FATAL_FAILURE(subscriber_.create_entities_xml(1, 0x80, UXR_STATUS_OK, 0));
+    }
+
+    void TearDown() override
+    {
+        ASSERT_NO_FATAL_FAILURE(publisher_.close_transport(transport_));
+        ASSERT_NO_FATAL_FAILURE(subscriber_.close_transport(transport_));
+    }
+
+    void init_agent(uint16_t port)
+    {
+        switch(transport_)
+        {
+            case UDP_TRANSPORT:
+                agent_.reset(new eprosima::uxr::UDPServer(port));
+                break;
+            case TCP_TRANSPORT:
+                agent_.reset(new eprosima::uxr::TCPServer(port));
+                break;
+        }
+        agent_->run();
+    }
+
+    void check_messages(std::string message, size_t number, uint8_t stream_id_raw)
+    {
+        std::thread publisher_thread(&Client::publish, &publisher_, 1, stream_id_raw, number, message);
+        std::thread subscriber_thread(&Client::subscribe, &subscriber_, 1, stream_id_raw, number, message);
+
+        publisher_thread.join();
+        subscriber_thread.join();
+    }
+
+protected:
+    int transport_;
+    std::unique_ptr<eprosima::uxr::Server> agent_;
+    Client publisher_;
+    Client subscriber_;
+    static const std::string SMALL_MESSAGE;
+};
+
+const std::string PublisherSubscriberInteraction::SMALL_MESSAGE("Hello DDS world!");
+
+INSTANTIATE_TEST_CASE_P(TransportAndLost, PublisherSubscriberInteraction,
+        ::testing::Combine(::testing::Values(UDP_TRANSPORT, TCP_TRANSPORT), ::testing::Values(0.0f, 0.05f, 0.1f)));
+
+TEST_P(PublisherSubscriberInteraction, PubSub1TopicsBestEffort)
+{
+    if(0.0f == std::get<1>(GetParam())) //only without lost
+    {
+        check_messages(SMALL_MESSAGE, 1, 0x01);
+    }
+}
+
+TEST_P(PublisherSubscriberInteraction, PubSub10TopicsBestEffort)
+{
+    if(0.0f == std::get<1>(GetParam())) //only without lost
+    {
+        check_messages(SMALL_MESSAGE, 10, 0x01);
+    }
+}
+
+TEST_P(PublisherSubscriberInteraction, PubSub1TopicsReliable)
+{
+    check_messages(SMALL_MESSAGE, 1, 0x80);
+}
+
+TEST_P(PublisherSubscriberInteraction, PubSub10TopicsReliable)
+{
+    check_messages(SMALL_MESSAGE, 10, 0x80);
+}
+
+/* Fix the non-reliable behavior when messages is higher than the agent history to enable this
```

Comment:
Please add a TODO with the associated task.

Reply:
Currently, there is not a task associated. Maybe it does not require a task and only a better interpretation of the apparent problem

Reply:
I added this task (4423).